src/js/legend.js

```diff
@@ -0,0 +1,227 @@
+            row: ".gpii-ca-pieChart-legend-table-row",
+            colorCell: ".gpii-ca-pieChart-legend-color-cell",
+            labelCell: ".gpii-ca-pieChart-legend-label-cell",
+            valueCell: ".gpii-ca-pieChart-legend-value-cell"
+        },
+        events: {
+            onLegendCreated: null  // Fire when the legend is created. Ready to register D3 DOM event listeners
+        },
+        listeners: {
+            "onCreate.create": {
+                funcName: "gpii.chartAuthoring.pieChart.legend.create",
+                args: ["{that}"]
+            }
+        },
+        modelListeners: {
+            dataSet: {
+                funcName: "{that}.draw",
+                excludeSource: "init"
+            }
+        },
+        invokers: {
+            draw: {
+                funcName: "gpii.chartAuthoring.pieChart.legend.draw",
+                args: ["{that}"]
+            },
+            sort: {
+                funcName: "gpii.chartAuthoring.pieChart.legend.sort",
+                args: ["{arguments}.0", "{arguments}.1"]
+            },
+            getColorCellStyle: {
+                funcName: "gpii.chartAuthoring.pieChart.legend.getColorCellStyle",
+                args: ["{arguments}.0"]
+            }
+        }
+    });
+
+    // Takes the dataSet array and the color array, and returns a consolidated object array to ease sorting and other operations while keeping colors "correct"
+    gpii.chartAuthoring.pieChart.legend.consolidateDataAndColors = function (model) {
+        var dataSet = model.dataSet;
+        var colors = (typeof(model.colors) === "function") ? model.colors : gpii.d3.getColorScale(model.colors);
+
+        var consolidated = [];
+        fluid.each(dataSet, function(item, i) {
+            var oneConsolidated = {
+                id: item.id,
+                label: item.label,
+                value: item.value,
+                color: colors(i)
+            };
+            consolidated.push(oneConsolidated);
+        });
+
+        return consolidated;
+
+    };
+
+    gpii.chartAuthoring.pieChart.legend.draw = function (that) {
+        var table = that.table,
+            legendOptions = that.options.legendOptions,
+            dataSet = that.model.dataSetWithColors,
+            rowClass = that.classes.row,
+            colorCellClass = that.classes.colorCell,
+            labelCellClass = that.classes.labelCell,
+            valueCellClass = that.classes.valueCell,
+            sort = legendOptions.sort;
+        var tbody = table.selectAll("tbody");
+
+        var rows = tbody.selectAll("tr")
+            .data(dataSet, function (d) {
+                return d.id;
+            });
+
+        // Add new rows for new data, apply appropriate classes for selectors and styling
+
+        var addedRows = rows.enter().append("tr");
+
+        addedRows
+        .attr({
+            "class": rowClass
+        });
+
+        addedRows
+        .append("td")
+        .attr({
```

Comment:
I just noticed that another issue is that the use of indentation isn't consistent within this function. We should pick a style and stick to it.